torchao/prototype/autoround/core.py

```diff
@@ -0,0 +1,391 @@
+    mod.to("cpu")
+    return out.cpu() if isinstance(out, torch.Tensor) else out
+
+
+@torch.no_grad()
+def infer_mod(mod, grouped_args, spec):
+    outputs = []
+    for i, inp in enumerate(grouped_args):
+        cur_args, cur_kwargs = tree_unflatten(inp, spec)
+        cur_kwargs.pop("idx")
+        out = _infer_mod(mod, cur_args, cur_kwargs)
+        outputs.append(out)
+    return outputs
+
+
+class _ModuleMapping:
+    def __init__(self):
+        self._modules_mapping: Dict[int, torch.nn.Module] = {}
+
+    def add_module(self, idx, mod):
+        self._modules_mapping[idx] = mod
+
+    def get_module(self, idx):
+        return self._modules_mapping[idx]
+
+    def reset(self):
+        self._modules_mapping = {}
+
+    def generate_id(self):
+        return len(self._modules_mapping) + 1
+
+
+module_mapping = _ModuleMapping()
+
+
+from torch.library import Library
+
+t_lib = Library("transformers_ops", "DEF")
+tran_ops = torch.ops.transformers_ops
+
+# The `general_decoder` serves as a flag to inform the dispatcher that this function(decoder block) is intended for optimization.
+# All of the args and kwargs will be passed to the optimized function,
+# that unpack them and return the correct output.
+# The call flow:
+#   `_DecoderLayerWrapper.forward`
+#       ->`general_decoder` under `__torch_function__`
+#       -> `optimize_decoder`
+#       -> return the optimized output
+t_lib.define("general_decoder(Tensor hidden_state) -> (Tensor, Tensor[])")
+
+
+def optimize_decoder(func, grouped_args, spec):
+    first_grouped_args = grouped_args[0]
+    first_cur_args, first_cur_kwargs = tree_unflatten(first_grouped_args, spec)
+    decoder_block_idx = first_cur_kwargs["idx"]
+    logging.info(f"Optimizing decoder layer {decoder_block_idx}")
+    decoder_block = module_mapping.get_module(decoder_block_idx)
+    origin_output = infer_mod(decoder_block, grouped_args, spec)
+    apply_auto_round(decoder_block, grouped_args, spec, origin_output)
+    return ar_utils.move_data_to_device(origin_output, "cpu")
+
+
+def _replace_buffers_and_params(model):
+    for name, buf in model.named_buffers(recurse=False):
+        setattr(model, name, MultiTensor([buf]))
+    for name, param in model.named_parameters(recurse=False):
+        setattr(model, name, torch.nn.Parameter(MultiTensor([param]), False))
+    return model
+
+
+def _revert_replace_buffers_and_params(model):
+    for name, buf in model.named_buffers(recurse=False):
+        if isinstance(buf, MultiTensor):
+            setattr(model, name, buf.values[0])
+    for name, param in model.named_parameters(recurse=False):
+        if isinstance(param, MultiTensor):
+            setattr(model, name, torch.nn.Parameter(param.values[0], False))
+    return model
+
+
+def _replace_with_custom_fn_if_matches_filter(
+    model: torch.nn.Module,
+    replacement_fn,
+    filter_fn,
+    cur_fqn="",
+) -> None:
+    if filter_fn(model, cur_fqn[:-1]):
+        model = replacement_fn(model)
```

Comment:
I noticed there is a similar function https://github.com/pytorch/ao/blob/934dead15981ae50b8ffc3f95a1fc9f02ac3af5c/torchao/quantization/quant_api.py#L151-L156

The main difference is that this function continues searching through the submodules of the replacement module. Should we extend the existing function to support that?

Reply:
I see, I feel it might be clearer to add this to utils and give it a different name, combining the code in one function might be confusing

Reply:
also I'm curious is this really needed? do you expect to replace a submodule and then replace its child?

Reply:
This function was copied from [multi_tensor_test.py](https://gist.github.com/HDCharles/a1b575bbf8875f994af8a01b225e1227#file-multi_tensor_test-py). In this case, rather than swapping a specific module with a new one, it replaces all buffers and parameters within the module with `MulTensor`. Consequently, it needs to iterate through the parameters of the module itself and those of all its submodules. A more straightforward implementation is below:

```python
import torch

def _update_model_and_children(model, replacement_fn):
    replacement_fn(model)
    for name, child in model.named_children():
        new_child = _update_model_and_children(child, replacement_fn)
        if new_child is not child:
            setattr(model, name, new_child)
    return model

def replace_buffers_and_params_with_zeros(model: torch.nn.Module):
    for name, buf in model.named_buffers(recurse=False):
        setattr(model, name, torch.zeros_like(buf))
    for name, param in model.named_parameters(recurse=False):
        setattr(model, name, torch.nn.Parameter(torch.zeros_like(param), False))
    return model

with torch.no_grad():
    class Mod(torch.nn.Module):
        def __init__(self):
            super().__init__()
            self.lin = torch.nn.Linear(10, 10)
            self.register_buffer("other", torch.randn(10, 20) * 0)

        def forward(self):
            pass

    model = Mod()
    _update_model_and_children(model, replace_buffers_and_params_with_zeros)
    for n, p in model.named_parameters():
        assert p.sum() == 0
```

Reply:
I see, do you need MultiTensor for buffers and params as well? I thought you only need it for input?

Reply:
Oh, you are right! Replacing input with `MultiTensor` is enough, thanks!